File: Android/shiba/src/main/java/moe/tlaster/shiba/scripting/visitors/JSViewVisitor.kt
package moe.tlaster.shiba.scripting.visitors

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.annotation.JsonAppend
import moe.tlaster.shiba.ShibaView
import moe.tlaster.shiba.common.Singleton
import moe.tlaster.shiba.scripting.conversion.toNative
import moe.tlaster.shiba.type.Property
import moe.tlaster.shiba.type.ShibaExtension
import moe.tlaster.shiba.visitors.ValueVisitor
import org.liquidplayer.javascript.JSObject
import org.liquidplayer.javascript.JSValue

internal object JSViewVisitor {
    private const val viewType = "IView"
    fun visit(value: JSValue) : Any? {
        if (!value.isObject) {
            return value.toNative()
        }
        val obj = value.toObject()
        if (!obj.hasProperty("className")) {
            return obj.toNative()
        }
        when (obj.property("className").toString()) {
            viewType -> {
                return visitView(obj)
            }
            else -> {
                throw IllegalArgumentException()
            }
        }
    }

    private fun visitView(value: JSObject): ShibaView {
        val name = value.property("name").toString()
        val properties = value.property("properties").toJSArray()
        val children = value.property("children").toJSArray()
        val view = ShibaView(name)
        properties.forEach {
            if (it is JSValue && it.isObject) {
                view.properties.add(visitProperty(it.toObject()))
            }
        }
        children.forEach {
            if (it is JSValue) {
                val child = visit(it)
                if (child is ShibaView) {
                    child.parent = view
                    view.children.add(child)
                } else {
                    view.defaultValue = child
                }
            }
        }
        return view
    }

    private fun visitProperty(value: JSObject): Property {
        val valueType = ValueType.fromInt(value.property("valueType").toNumber().toInt())
        val name = value.property("name").toString()
        val propertyValue = value.property("value")
        when (valueType) {
            ValueType.Extension -> {
                val extValue = propertyValue.toObject()
                val target = extValue.property("target").toString()
                val type = extValue.property("type").toString()
                val scriptName = if (extValue.hasProperty("scriptName")) {
                    extValue.property("scriptName").toString()
                } else {
                    null
                }
                return Property(name, ShibaExtension(type, target, scriptName))
            }
            ValueType.Boolean, ValueType.Number, ValueType.String, ValueType.Null -> {
                return Property(name, propertyValue.toNative())
            }
            ValueType.Custom -> {
                return Property(name, ValueVisitor.visit(Singleton.get<ObjectMapper>().readTree(propertyValue.toString()), null))
            }
            null -> {
                return Property(name, null)
            }
        }
    }
}
internal enum class ValueType(val value: Int) {
    Extension(0),
    Boolean(1),
    Number(2),
    String(3),
    Null(4),
    Custom(5);
    companion object {
        private val values = values()
        fun fromInt(value: Int) = values.firstOrNull { it.value == value }
    }
}

File: UWP/Shiba/Scripting/Visitors/JSViewVisitor.cs
using System;
using System.Collections.Generic;
using ChakraHosting;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Shiba.Controls;
using Shiba.Internal;
using Shiba.Visitors;

namespace Shiba.Scripting.Visitors
{
    internal class JSViewVisitor
    {
        private const string ViewType = "IView";

        public object Visit(JavaScriptValue value)
        {
            const string className = "className";
            if (value.ValueType != JavaScriptValueType.Object || !value.HasProperty(className.ToJavaScriptPropertyId()))
                return value.ToNative();
            var type = value.GetProperty(className.ToJavaScriptPropertyId()).ToNative<string>();
            switch (type)
            {
                case ViewType:
                    return VisitView(value);
                default: throw new ArgumentOutOfRangeException();

            }
        }

        private View VisitView(JavaScriptValue value)
        {
            var name = value.GetProperty("name".ToJavaScriptPropertyId()).ToNative<string>();
            var properties = value.GetProperty("properties".ToJavaScriptPropertyId());
            var children = value.GetProperty("children".ToJavaScriptPropertyId());
            var view = new View(name, string.Empty);
            if (properties.ValueType != JavaScriptValueType.Array)
            {
                throw new ArgumentException();
            }

            if (children.ValueType != JavaScriptValueType.Array)
            {
                throw new ArgumentException();
            }

            {
                var currentIndex = 0;
                while (properties.HasIndexedProperty(currentIndex.ToJavaScriptValue()))
                {
                    view.Properties.Add(VisitProperty(properties.GetIndexedProperty(currentIndex.ToJavaScriptValue())));
                    currentIndex++;
                }
            }

            {
                var currentIndex = 0;
                while (children.HasIndexedProperty(currentIndex.ToJavaScriptValue()))
                {
                    var child = Visit(children.GetIndexedProperty(currentIndex.ToJavaScriptValue()));
                    if (child is View childView)
                    {
                        childView.Parent = view;
                        view.Children.Add(childView);
                    }
                    else
                    {
                        view.DefaultValue = child;
                    }
                    currentIndex++;
                }
            }

            return view;
        }

        private Property VisitProperty(JavaScriptValue value)
        {
            Enum.TryParse<ValueType>(value.GetProperty("valueType".ToJavaScriptPropertyId()).ToNative() + "", out var valueType);
            var name = value.GetProperty("name".ToJavaScriptPropertyId()).ToNative<string>();
            var propertyValue = value.GetProperty("value".ToJavaScriptPropertyId());
            switch (valueType)
            {
                case ValueType.Extension:
                    var target = propertyValue.GetProperty("target".ToJavaScriptPropertyId()).ToNative<string>();
                    var type = propertyValue.GetProperty("type".ToJavaScriptPropertyId()).ToNative<string>();
                    var scriptName = string.Empty;
                    if (propertyValue.HasProperty("scriptName".ToJavaScriptPropertyId()))
                    {
                        scriptName = propertyValue.GetProperty("scriptName".ToJavaScriptPropertyId()).ToNative<string>();
                    }
                    return new Property(name, new ShibaExtension(type, target, scriptName));
                case ValueType.Custom:
                    return new Property(name, Singleton<ValueVisitor>.Instance.DynamicVisit(propertyValue, null));
                case ValueType.Boolean:
                case ValueType.Number:
                case ValueType.String:
                case ValueType.Null:
                    return new Property(name, propertyValue.ToNative());
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }

        private enum ValueType
        {
            Extension,
            Boolean,
            Number,
            String,
            Null,
            Custom
        }
    }
}